sdk/identity/azure-identity/azure/identity/_credentials/managed_identity.py

```diff
@@ -96,12 +96,14 @@ def __init__(self, **kwargs: Any) -> None:
             _LOGGER.info("%s will use IMDS", self.__class__.__name__)
             self._credential = ImdsCredential(**kwargs)
 
-    def __enter__(self):
-        self._credential.__enter__()
+    def __enter__(self) -> "ManagedIdentityCredential":
+        if self._credential:
+            self._credential.__enter__()  # type: ignore
```

Comment:
Do all sync credential types define an __enter__ and __exit__? I'm wondering why the TokenCredential protocol doesn't define them... (assuming that's why we had to ignore here). We do include them on the async protocol by inheriting AsyncContextManager.

sync: https://github.com/Azure/azure-sdk-for-python/blob/main/sdk/core/azure-core/azure/core/credentials.py#L22
async: https://github.com/Azure/azure-sdk-for-python/blob/main/sdk/core/azure-core/azure/core/credentials_async.py#L13

Reply:
The problem here is TokenCredential is a GAed protocol and if we add enter & exit, seems like it will be a breaking change.

Reply:
discussed offline